### PR TITLE
633/workflow entryname

### PR DESCRIPTION
**Pull request by Richard-Hansen**
Github Issue: dockstore/dockstore#2309

The user can specify a custom name for a workflow with the '--entryname' parameter when publishing the workflow from the CLI. This will create a separate workflow entry from the default registered workflow.

### CODE REVIEW

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: GitHub doesn't let me comment on lines you haven't changed, but I'm reading the whole method to better understand what is going on, and noticed this: on line 650/651 the error message is saying it can't publish/unpublish `newName`, but it doesn't even use `newName` on the previous line. I think you want that to be `entryPath`.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:668]: This method is already invoked on line 648. Can't you just use `existingWorkflow`? Note, this is a network request, a relatively expensive call, so you don't want to duplicate this call if unnecessary.

Compiler may be possibly be giving you a variable undefined error, if so, but you should fix the code.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:682]: Is this line necessary? I think it is, but just double-checking.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: This was also like this, but for the block starting on line 652/653, I feel that we should have a message saying `newName` is not valid for unpublish.

**Review by coverbeck — CHANGES_REQUESTED** [review]

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: The current implementation doesn't handle publishing/unpublishing workflows with '--entryname' gracefully if there is an error. For example, if I try to publish a workflow with a custom name that is already published, the function will throw an ApiException instead of printing a clean "This workflow is already published" message.

I'll fix this in my next commit, but would like to clarify expected functionality:
`dockstore workflow publish --unpub --entry workflowpath --entryname newname` 
vs 
`dockstore workflow publish --unpub --entry workflowpath/entryname`

Currently, in the first case, the entry name is ignored and the command is interpreted as: 
`dockstore workflow publish --unpub --entry workflowpath`.
 Would it make sense to have the two cases listed above be logically equivalent in execution? Or should the function prevent use of --unpub with --entryname?

Comment (coverbeck) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: I will defer to @garyluu and @denis-yuen , but here are my thoughts:

1. If we're going to support `--entryname` we should add it to the [help message](https://github.com/dockstore/dockstore-cli/blob/5de67a497465612ddc693ffc694e916eea018af2/dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java#L689). I didn't realize until reading the issue again that this an enhancement.
2. Depending on how you want to approach it, either that help message should say that the argument is ignored for unpublish; or, the help message says it's only applicable to publish and if we detect it here for unpublish, we throw an error.

In either case, I think this renders your question moot; --entryname should either be ignored or cause an error if used with unpublish.

In hindsight, I wished we had called it something like `newEntryName`.

Comment (garyluu) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: I'm personally not a fan of flags that conflict with other flags when it can be helped. I'd rather there be an unpublish command. However, in order to keep things similar to tools, I suppose that shouldn't be done.

Anyways, given the current situation, I think I agree with Charles. What does tools do right now?

Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: Tools functions in a [similar way](https://github.com/dockstore/dockstore-cli/blob/5de67a497465612ddc693ffc694e916eea018af2/dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java#L213) to workflows. The `--entryname` parameter is ignored for unpublish, regardless of if it is set or not. Both could use slightly more descriptive error messages for when publish/unpublish fails.

I can modify workflows to better handle `--entryname` with unpublish and provide more descriptive errors. Should i do the same for tools/create a similar ticket for tools? Furthermore, I think @coverbeck's idea to change `--entryname` to something such as `--newEntryName` would be beneficial.

Comment (coverbeck) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: The problem is, even though it's extremely unlikely, is that there could be some script somewhere using `--entryname`, and changing the parameter name would break that script. Some ideas:

1. Only make it `newEntryName` for workflows, since it's a new feature there. The downside is that it's asymmetrical with tools though.
2. Make `--newEntryName` an additional option for tools. Continue to support `--entryname`, but undocumented, i.e., the help message for tools only lists `--newEntryName`.
3. It's unlikely anybody is using this feature, just change it on tools.

I suggest the first option for now, and wait for Denis' opinion.

Comment (garyluu) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: I think if you have time to make all the changes (for both tools and workflows) mentioned by Charles then go ahead. Otherwise, I think you should first match the functionality of tools and workflows.   And then create tickets to simultaneously make improvements to both.

Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:667]: I'll update both tools and workflows to have the same handling of the `--entryname` parameter, to ensure they are logically symmetric. I'll move forward from there after getting Denis' opinion.

Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:676]: Updated. Now reports that the error occurred when trying to locate the registered workflow at `entryPath`.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java]: Updated. No longer makes the additional unnecessary API call.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:691]: Ideally, manualRegister will not return a null value, and throw the expected API exception. In the theoretical case it doesn't, this is extra verbose.

**Review by coverbeck — APPROVED** [review]
LGTM to me, but pending decision on entryname when Denis gets back.

**Review by garyluu — APPROVED** [review]

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (denis-yuen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:676]: This looks like its going into develop ... I'm ok with 2. actually (rename to better parameter name but allow old name for backwards compatibility). Also ok with a series of PRs for the same reason (I've be more nervous about a partial incomplete fix for 1.9.x)

Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:676]: Ok, I'll update `--entryName` to `--newEntryName`, and modify the help messages to specify `--newEntryName` but still keep support for `--entryName`.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java:458]: 1. Perhaps a warning that `--entryName` is deprecated? 
2. Also, checking for it here, you're going to be supporting --entryName for workflows. I thought the idea was tool would support --entryName and --newEntryName, because tool already has --entryName, but since it's being added to workflow for the first time, we don't have to worry about backwards compatibility.

I don't feel too strongly about either of these.

File: dockstore-cli-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java
Comment (coverbeck) [inline dockstore-cli-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java:1083]: Would be good to have tests that verify both --newEntryName and --entryName work (all these tests are for workflow).

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java:227]: I don't think this if test is necessary. If line 221 fails to find a tool, it will throw an exception, and line 224 will exit. And if it didn't throw an exception, you'd get an NPE on line 222.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:658]: Same as other comment; I don't think you need this block.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:689]: Won't newWorkflow always be set? If line 678 failed, won't it throw an exception?

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java:458]: 1. Adding a deprecation warning
2. My initial thinking was that it would seem more consistent to have `--entryname` compatible, but deprecated, for both tools and workflows. Since this is currently (happy to change it) being handled within the AbstractEntryClient, I opted to avoid specifying tools/workflows specifically. Thoughts?

File: dockstore-cli-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java
Comment (Richard-Hansen) [reply dockstore-cli-integration-testing/src/test/java/io/dockstore/client/cli/GeneralWorkflowIT.java:1083]: Adding tests for both `--entryname` and `--newEntryName` for tools and workflows.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/ToolClient.java:227]: I see what you're saying, thanks for the catch! Removing the unnecessary lines.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:658]: Agreed. Same as above.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java:689]: Ideally, yes. I added this because if, in some strange case, the `manualRegister()` API call had null as its return value, that would not throw an exception from the register function itself (We would hit a null pointer exception 2 lines later, so it wouldn't make it very far).

Either way, it is redundant, I will remove it from both tools and workflows.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java
Comment (coverbeck) [inline dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java:457]: Unless we already do this elsewhere, I don't think we add mixed case parameter names to the command-line.

**Review by coverbeck — COMMENTED** [review]
Looks good, but I'd missed the mixed case parameter the first time through. Even we lower-case the args in the code (do we?) before testing, we should keep the same patterns throughout.

File: dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java
Comment (Richard-Hansen) [reply dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java:457]: Looks like dash-delimited is the most common, `word-word-word`. I'll switch to that pattern. Thanks!

**Comment by coverbeck** [comment]
> This pull request **introduces 2 alerts** when merging [85ee68e](https://github.com/dockstore/dockstore-cli/commit/85ee68e1a3818df1e5cf5b1f94f5f0847d7d3b8e) into [4a9474a](https://github.com/dockstore/dockstore-cli/commit/4a9474a7eae4e5d38f317a102e57593510d002af) - [view on LGTM.com](https://lgtm.com/projects/g/dockstore/dockstore-cli/rev/pr-0c5438b436f20169e385eb30968ba3cdcc6e10d3)
> 
> **new alerts:**
> 
>     * 2 for Dereferenced variable may be null

Sorry, looks like that was my fault. :(

**Review by denis-yuen — APPROVED** [review]

**Review by garyluu — APPROVED** [review]

**Pull request merged**
